[serve] Update docs for v2 Deployments API

edoakes opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number

Closes https://github.com/ray-project/ray/issues/14809
Closes https://github.com/ray-project/ray/issues/14808

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

edoakes commented:
@architkulkarni @simon-mo PTAL, this is still somewhat rough but I've gotten up through the "Core APIs' page (going sequentially through the existing pages). There are a few places where I have TODOs to put in links that don't exist yet.

architkulkarni reviewed (commented):
I wrote some nits.  Looks great!

edoakes commented on python/ray/serve/examples/doc/imported_backend.py:
```diff
@@ -1,11 +1,13 @@
-serve.create_endpoint("imported", backend="imported", route="/imported")
 
-print(requests.get("http://127.0.0.1:8000/imported").text)
+MyDeployment.deploy("/model_path.pkl")
```
updated the args to be more clear

edoakes commented on doc/source/serve/core-apis.rst:
```diff
@@ -4,109 +4,124 @@ Core APIs
 
-We can also query the endpoint using the :mod:`ServeHandle <ray.serve.handle.RayServeHandle>` interface.
+
+If you want to avoid always re-deploying, this behavior can be controlled by providing a ``version`` string for the deployment in the decorator or ``Deployment.options()``.
+If a ``version`` string is provided, a rolling update of the replicas will *only* happen if the version is updated.
+Otherwise, the call to ``.deploy()`` will be a no-op.
+This can be useful when you have many deployments in a script and you only want to re-deploy one of them each time the script runs.
```
Thanks :)

edoakes commented:
Ok @architkulkarni @simon-mo I updated the rest of the docs, this should be ready for a full review.

@architkulkarni I updated `core-apis.rst` with your suggested changes if you want to skim through it quickly again.

simon-mo commented on doc/source/serve/core-apis.rst:
```diff
@@ -1,128 +1,150 @@
-By default endpoints are serviced only by the backend provided to :mod:`create_endpoint <ray.serve.api.create_endpoint>`, but in some cases you may want to specify multiple backends for an endpoint, e.g., for A/B testing or incremental rollout.
-For information on how to do this, please see :ref:`serve-split-traffic`.
+  >> serve.list_deployments()
+  {'MyFirstDeployment': <ray.serve.api.ServeDeployment object at 0x7fe0c04c42b0>}
```
oof, we need a better `__repr__` for ServeDeployment

edoakes replied:
Added! There was already a __str__, I just called into it for repr

simon-mo commented on doc/source/serve/core-apis.rst:
```diff
@@ -1,128 +1,150 @@
 
-  >>> serve.list_endpoints()
-  {'simple_endpoint': {'route': '/simple', 'methods': ['GET'], 'traffic': {}}}
+  @serve.deployment(name="my_deployment", num_replicas=1)
```
add comment this is the SAME file over time

edoakes replied:
done

simon-mo commented on doc/source/serve/http-servehandle.rst:
```diff
@@ -1,92 +1,142 @@
+    class FastAPIWrapper:
+        pass
+
+This is useful for scaling out an existing FastAPI app with no modifications necessary. You can also combine routes defined this way with routes defined on the Deployment:
```
add a note saying existing middlewares, docs, and advanced feature, etc should work as-is

simon-mo commented on doc/source/serve/http-servehandle.rst:
````diff
@@ -110,146 +160,85 @@ You can follow the same pattern for other Starlette middlewares.
-
-When the request arrives in the model, you can access the data similarly to how
-you would with an HTTP request. Here are some examples how Ray Serve's built-in 
-``ServeRequest`` mirrors ```starlette.requests.request``:
````
did we end up deleting `ServeRequest`?

edoakes replied:
Deployments get regular args and kwargs now

architkulkarni commented on doc/source/serve/tutorials/batch.rst:
```diff
@@ -88,12 +88,11 @@ Ray Serve was able to evaluate them in batches.
+or the Python API. Learn more :ref:`here<serve-handle-explainer>`.
 
-To query the backend via Python API, we can use ``serve.get_handle`` to receive
+To query the backend via Python API, we can use ``Deployment.get_handle`` to receive
```
```suggestion
To query the deployment via the Python API, we can use ``Deployment.get_handle`` to receive
```

architkulkarni commented on python/ray/serve/examples/doc/aiohttp/aiohttp_deploy_serve.py:
```diff
@@ -9,13 +9,12 @@
 
-# Define a function to serve. Alternatively, you could define a stateful class.
+# Set up a deployment with the desired number of replicas. This could also be
+# e a stateful class (e.g., if we had an expensive model to set up).
```
```suggestion
# a stateful class (e.g., if we had an expensive model to set up).
```

architkulkarni commented on python/ray/serve/examples/doc/fastapi/servehandle_fastapi.py:
```diff
@@ -6,32 +6,28 @@
 @app.get("/generate")
 async def generate(query: str):
-    return await serve_handle.remote(query)
+    # Get a handle to our Ray Serve endpoint so we can query it in Python.
```
```suggestion
    # Get a handle to our Ray Serve deployment so we can query it in Python.
```

architkulkarni commented on python/ray/serve/examples/doc/fastapi/servehandle_fastapi.py:
```diff
@@ -6,32 +6,28 @@
 
 app = FastAPI()
```
Is it worth keeping this "Scaling up a FastAPI application" section around now that we have the tighter FastAPI integration?

edoakes replied:
Good question. I think it's fine to keep for now but we might consider updating it to use the new "drop-in" support

architkulkarni reviewed (approved):
Looks great!  Added a few more nits.

edoakes commented on doc/source/serve/http-servehandle.rst:
```diff
@@ -1,92 +1,142 @@
-the ``username`` parameter will be accessible in your backend code as follows:
+    MyFastAPIDeployment.deploy()
+
+Now if you send a request to ``/hello``, this will be routed to the ``root`` method of our Deployment. We can also easily leverage FastAPI to define multiple routes with different HTTP methods:
```
yeah we should be consistent. I'll make them all lower case.
